### PR TITLE
Performance boost

### PR DESCRIPTION
**Pull request by knrz**
I benchmarked by providing the following sample to the functions.

```elixir
sample = 1..100_000 |> Enum.shuffle |> Enum.take(50_000)
```

### CODE REVIEW

File: lib/statistics.ex [hunk 1]
```diff
@@ -18,21 +10,30 @@ defmodule Statistics do
 
   Calls Enum.sum/1
   """
-  def sum(list) do
-    Enum.sum(list)
-  end
+  def sum([]), do: nil
+  def sum(list), do: sum(list, 0)
+  def sum([x|xs], sum), do: sum(xs, sum + x)
```
Comment (msharp) [inline lib/statistics.ex]: Do you think the sum/2 functions should be private?

File: lib/statistics.ex [hunk 2]
```diff
@@ -18,21 +10,30 @@ defmodule Statistics do
...
+
       iex> Statistics.mean([1,2,3])
       2.0
 
   """
-  def mean(list) do
-    Enum.sum(list) / Enum.count(list)
+  def mean([]), do: nil
+  def mean(list), do: mean(list, 0, 0)
+  def mean([x|xs], length, sum) do
```
Comment (msharp) [inline lib/statistics.ex]: Also, I feel that these mean/3 methods should be private.

File: lib/statistics.ex (hunk 1 above)
Comment (knrz) [inline lib/statistics.ex]: The alternative is:

```elixir
def sum(list) when is_list(list), do: _sum(list, 0)

defp _sum([], _), do: 0
defp _sum([x|xs], n), do: sum(xs, sum + x)
```

On second thought, I like the above more. Your call though.

File: lib/statistics.ex (hunk 2 above)
Comment (knrz) [inline lib/statistics.ex]:
```elixir
def mean(list) when is_list(list), do: _mean(list, 0, 0)

defp _mean([], 0, 0), do: nil
defp _mean([], length, sum), do: sum / length
defp _mean([x|xs], length, sum), do: _mean(xs, length + 1, sum + x)
```

Like that?

**Comment by knrz** [comment]
I messed up the git history when working on this. Feel free to close the issue and I'll implement it in a new, less convoluted PR.

File: lib/statistics.ex (hunk 1 above)
Comment (msharp) [inline lib/statistics.ex]: This is better but I probably prefer the `defp do_function` style for these recursive implementations which are not exported by the module.

I think the `_function` naming scheme is coming from Python (where nothing is private - so an underscore prefix is a convention that signals that a function is not intended to be part of the API)

Comment (knrz) [inline lib/statistics.ex]: Also, `_variable` is a way of annotating unused variables in Elixir—`_function` might be an extension of that. Anyway, I'm closing this PR and coming up with a cleaner one.

**Pull request closed without merging**